src/core/ucc_context.c

```diff
@@ -204,3 +205,84 @@ void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
             config->lib->full_prefix, (ucc_config_print_flags_t)flags);
     }
 }
+
+ucc_status_t ucc_context_create(ucc_lib_h lib,
+                                const ucc_context_params_t *params,
+                                const ucc_context_config_h  config,
+                                ucc_context_h *context)
+{
+    ucc_cl_context_params_t cl_ctx_params;
+    ucc_cl_context_t       *cl_ctx;
+    ucc_cl_lib_t           *cl_lib;
+    ucc_context_t          *ctx;
+    ucc_status_t            status;
+    uint64_t                i;
+    int                     num_cls;
+
+    num_cls = config->n_cl_cfg;
+    ctx     = ucc_malloc(sizeof(ucc_context_t), "ucc_context");
+    if (!ctx) {
+        ucc_error("failed to allocate %zd bytes for ucc_context",
+                  sizeof(ucc_context_t));
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    ctx->lib = lib;
+    memcpy(&ctx->params, params, sizeof(ucc_context_params_t));
```

Comment:
This may cause an ABI issue: if an application was built with an older UCC than the one it's running with, then this memcpy will might access a bigger ucc_context_params_t than the application gives - potentially causing a segfault (unlikely). I think a 100% future-proof solution would be to detect the memcpy size based on the field mask in params.

Reply:
@alex--m by "detect based on field mask" - do you mean copying fields one-by-one checking if the bit is set?

Reply:
I did something similar in xUCG: https://github.com/openucx/xucg/blob/531d0a7a40ab2775ba0d9123ba9acb41140ff60a/base/ucg_context.c#L226 

Not sure it's very elegant, but it gets the job done (and only calls memcpy once), and in retrospect maybe a per-field copy is actually the right thing to do here.

Reply:
Given that this is prevalent in UCC, it might be worth to create a copy routine. Between, how does UCX handle this?

Reply:
UCX doesn't copy params at any point. It goes field-by-field and sets internal variables/structures based on params (not calling memcpy).